crates/sui-indexer/src/models/events.rs

```diff
@@ -77,96 +88,138 @@ pub fn read_last_event(pg_pool_conn: &mut PgPoolConnection) -> Result<Option<Eve
     })
 }
 
-// NOTE: no need to retry here b/c errors here are not transient,
-// instead we write them to PG tables for debugging purposes.
-pub fn event_to_new_event(e: SuiEventEnvelope) -> Result<NewEvent, IndexerError> {
-    let event_json = serde_json::to_string(&e.event).map_err(|err| {
+pub fn commit_events(
+    pg_pool_conn: &mut PgPoolConnection,
+    mut event_envelopes: Vec<IndexerEventEnvelope>,
+    txn_page_next_cursor: Option<TransactionDigest>,
+) -> Result<Option<(usize, TransactionDigest)>, IndexerError> {
+    // No op when there is no more than 1 event envelope,
+    // which could have  been left as next cursor from last iteration.
+    if event_envelopes.len() <= 1 {
+        return Ok(None);
+    }
```

Comment:
Skipping if `len() <= 1` (as opposed to `len() == 0`) is a bit hard to reason about here. Better to move it out of this function if the intent is to skip committing under some conditions

Reply:
thanks for the review, this is not pretty solution indeed!

when there is only one txn here, it indicates that the next_cursor is also None, so in that sense, if we commit this txn, then no txn can be used as the next cursor. If we lost track of next cursor, the best the indexer can do is to use the last txn, in that sense, one txn will be duplicate. we can count on table to do de-dup, but that's unnecessary DB cost I wanted to get rid of.

Ideally it would be easier if pagination API can always guarantee a Non-null next cursor (meaning that if FN has 100 txns in total, this API only returns 99 and leave the last as the cursor), b/c today a cursor of None could mean 1) start from scratch while reading a new txn page 2) no more txns are available on FN
which is indexer tries to deal with here.

Another approach could be, on the page fetching API side, we add an option to specify if we want to include / exclude the next cursor, such that we can use the last digest as the excluded cursor.

We technically do not have to change the existing JSON RPC endpoints, but for the other FN <> indexer interface (gRPC?), would be great to consider this.  cc @patrickkuo